### PR TITLE
no more infinite loops

### PR DESCRIPTION
**Pull request by Broshen**
- still crashes sometimes
- always finishes with unfreed memory

### CODE REVIEW

File: student.cc [hunk 1]
```diff
@@ -16,37 +16,41 @@ void Student::main() {
...
 				try {
 					machine -> buy((VendingMachine::Flavours)favFlavour, *giftCard());
 					prt.print(Printer::Student, id, 'G', favFlavour, giftCard() -> getBalance());
+					toPurchase--;
 					giftCard.reset();
 					break;
 				} catch (VendingMachine::Free &) {
 					yield(4);
-					i--;
+					toPurchase--;
```
Comment (jiayi-zhou) [inline student.cc]: free drinks do not decrement toPurchase

File: student.cc [hunk 2]
```diff
@@ -16,37 +16,41 @@ void Student::main() {
...
 			} or _Select(watcard) {
 				try {
 					machine -> buy((VendingMachine::Flavours)favFlavour, *watcard());
 					prt.print(Printer::Student, id, 'B', favFlavour, watcard() -> getBalance());
+					toPurchase--;
 					break;
 				} catch (VendingMachine::Free &) {
 					yield(4);
-					i--;
+					toPurchase--;
```
Comment (jiayi-zhou) [inline student.cc]: same as above

File: watcard_office.h [hunk 3]
```diff
@@ -25,6 +25,7 @@ _Task WATCardOffice {
 		Printer &prt;
 		void main();
 	public:
+		bool done=false;
```
Comment (jiayi-zhou) [inline watcard_office.h:28]: should this be public?

Comment (Broshen) [reply watcard_office.h:28]: yes, otherwise WATCardOffice can't set it to true on destructor

File: watcard_office.cc [hunk 4]
```diff
@@ -46,31 +26,85 @@ void WATCardOffice::main() {
...
+	prt.print(Printer::WATCardOffice, 'W');
 	return nextJob;
 }
 
 WATCardOffice::~WATCardOffice() {
+	for (unsigned int i = 0; i < numCouriers; i ++) {
+		courierPool[i]->done = true;
+	}
+	for (unsigned int i = 0; i < numCouriers; i ++) {
+		jobReady.signalBlock();
```
Comment (Broshen) [inline watcard_office.cc:70]: note: try signal()

File: student.cc [hunk 5]
```diff
@@ -16,37 +16,41 @@ void Student::main() {
...
 					break;
 				} catch (VendingMachine::Stock &) {
-					 machine = nameServer.getMachine(id);
-					 break;
+					machine = nameServer.getMachine(id);
+					prt.print(Printer::Student, id, 'V', machine -> getId());
+					break;
 				} catch (VendingMachine::Funds &) {
 					cardOffice.transfer(id, machine -> cost() + 5, watcard());
 					break;
```
Comment (Broshen) [inline student.cc:54]: i _think_ i should wait here for funds to complete transfer instead of breaking & yielding?

**Pull request merged**